### PR TITLE
WIP: Implementation proposition

### PR DESCRIPTION
**Pull request by mateusz-kolecki**
Hi!
I would like to help create Command Bus library for TypeScript. Any feedback will be appreciated.

### CODE REVIEW

File: src/commandBus.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+
+export interface Command { } // marking interface
+
+export interface Handler<C extends Command> {
+    handle(command: C): Promise<any>;
+}
+
+export interface CommandBus {
+    handle(command: Command): Promise<any>
```
Comment (snapshotpl) [inline src/commandBus.ts]: void

File: src/middlewares/commandLogger.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    private errorLevel: string;
+
+    constructor(logger: CommandLogger, options: CommandLoggerOptions = {}) {
+        this.logger = logger;
+        this.level = options.level || 'info';
+        this.errorLevel = options.errorLevel || 'error';
+    }
+
+    public async run(command: Command, next: NextMiddleware): Promise<any> {
+        const start = Date.now();
```
Comment (snapshotpl) [inline src/middlewares/commandLogger.ts]: Move before call `next()`

File: src/middlewares/commandLogger.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        this.logger.log(this.level, 'Command %s dispatched', commandName);
+
+        try {
+            return await next();
+        } catch (err) {
+            this.logger.log(this.errorLevel, 'Command %s error. %s', commandName, err);
+            throw err;
+        } finally {
+            const time = Date.now() - start;
+            this.logger.log(this.level, 'Command %s time %dms', commandName, time);
```
Comment (snapshotpl) [inline src/middlewares/commandLogger.ts]: Do you want to log command two times if error occured?

File: src/middlewares/index.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+
+import { Command } from '../commandBus';
+
+export type NextMiddleware = () => Promise<any>;
```
Comment (snapshotpl) [inline src/middlewares/index.ts]: Next middleware should receive command as arg

File: src/resolvers/memory.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    public register<T>(commandConstructor: ConstructorOf<T>, handler: Handler<T>): HandlerResolverInMemory {
+        this.handlers.set(commandConstructor, handler);
+        return this;
+    }
+
+    public resolve<T>(command: Command): Handler<T> {
+        const handler = this.handlers.get(command.constructor as ConstructorOf<Command>);
+
+        if (!handler) {
+            throw new Error(`Could not resolve handler for ${command.constructor.name}`);
```
Comment (snapshotpl) [inline src/resolvers/memory.ts:19]: imo this exception should come from `get()`

File: src/resolvers/memory.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        }
+
+        return handler;
+    }
+
+    public count(): number {
```
Comment (snapshotpl) [inline src/resolvers/memory.ts]: Are we need this?

File: src/resolvers/memory.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        return this.handlers.size;
+    }
+
+    public entries(): IterableIterator<[ConstructorOf<Command>, Handler<Command>]> {
```
Comment (snapshotpl) [inline src/resolvers/memory.ts]: Are we need this?

File: src/resolvers/memory.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+
+import { Handler, Command } from '../commandBus';
+import { HandlerResolver } from './index';
+
+export type ConstructorOf<T> = new (...args: any[]) => T;
+
+export class HandlerResolverInMemory implements HandlerResolver {
+    private handlers = new Map<ConstructorOf<Command>, Handler<Command>>();
+
+    public register<T>(commandConstructor: ConstructorOf<T>, handler: Handler<T>): HandlerResolverInMemory {
```
Comment (snapshotpl) [inline src/resolvers/memory.ts:10]: Why not to get this list by constructor?

File: test/component/commander.test.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    it("should resolve with value from handler", async () => {
+        const promise = commander.handle(new HelloCommand('John', 'Doe'));
+
+        return expect(promise).to.eventually.be.equal('Hello, John Doe');
+    });
+
+    it("should write info logs", async () => {
+        await commander.handle(new HelloCommand('John', 'Doe'));
+
+        const infos = logger.messages.filter(entry => entry.level === 'info');
```
Comment (snapshotpl) [inline test/component/commander.test.ts]: Are we want to test logger middeware here?

File: test/fakes/fakeLogger.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
+
+import { format } from 'util';
+import { CommandLogger } from '../../src/middlewares';
+
+
+export class FakeLogger implements CommandLogger {
+
+    public messages: { level: string, msg: string, meta: any }[] = [];
```
Comment (snapshotpl) [inline test/fakes/fakeLogger.ts]: if public, can be changed out of scope

File: test/unit/commander.test.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
+
+import { expect } from "chai";
+
+import { Commander } from '../../src/commander';
+import { Command } from '../../src/commandBus';
+import { Middleware, NextMiddleware } from '../../src/middlewares';
+
+
+class CallbackMiddleware implements Middleware {
```
Comment (snapshotpl) [inline test/unit/commander.test.ts:9]: final

Comment (snapshotpl) [reply test/unit/commander.test.ts:9]: Can exist as provided implementation

Comment (mateusz-kolecki) [reply test/unit/commander.test.ts:9]: TypeScript do not offer `final` classes :(

And this class is only for testing.

File: test/fakes/fakeLogger.ts (hunk 10 above)
Comment (mateusz-kolecki) [reply test/fakes/fakeLogger.ts]: This is also dump implementation for testing. I can use read-only getter like this:

```typescript
class FakeLogger implements CommandLogger {
    // ...
    public get Messages() {
        return this.messages;
    }
}
```

and then in test code:

```typescript
const infos = logger.Messages.filter(...);
```

What do you think about this?

File: src/resolvers/memory.ts (hunk 8 above)
Comment (mateusz-kolecki) [reply src/resolvers/memory.ts:10]: I wanted to be sure that each entry in `Map` is a command constructor as key and handler instance which handles exacly this particular command as value.
Now it can be checked at compile time for `register()` method calls.
Have no idea, how to declare `handlers` property to get same result.
Now propert is declared as any command constructor as value and any handler instane which handles any command.

In other words each call to `register()` method is made with some generic type which is inferred from command constructor type and this allow compiler to check that handler instance matches that command type.

This prevent mistakes like registering handler for another command.

File: src/resolvers/memory.ts (hunk 5 above)
Comment (mateusz-kolecki) [reply src/resolvers/memory.ts:19]: Can you explain in more details?

File: src/middlewares/index.ts (hunk 4 above)
Comment (mateusz-kolecki) [reply src/middlewares/index.ts]: Hmm... this was something I was thinking about for a long time...

My conclusion was this: if `next()` callback would receive command as argument then for most of the time it would be the same command in all meddlewares across the chain invocation. In other words it would be the same command as this passed to `run()` method.
Now i pass command to next middleware in the background (see for execution chain reducer);

Ok, but maby middleware can change command? Maby middleware creator would create his own command instance and pass this new one to `next()`?
In my opinion sneaky...

What do You think about this?

File: test/fakes/fakeLogger.ts (hunk 10 above)
Comment (mateusz-kolecki) [reply test/fakes/fakeLogger.ts]: done

File: src/middlewares/commandLogger.ts (hunk 3 above)
Comment (mateusz-kolecki) [reply src/middlewares/commandLogger.ts]: done.

File: src/middlewares/index.ts (hunk 4 above)
Comment (mateusz-kolecki) [reply src/middlewares/index.ts]: ok, i added command as argument

Comment (mateusz-kolecki) [reply src/middlewares/index.ts]: done

File: src/resolvers/memory.ts (hunk 5 above)
Comment (mateusz-kolecki) [reply src/resolvers/memory.ts:19]: closed
